### PR TITLE
[FLINK-16480][e2e] On failure, add e2e logs into artifact

### PR DESCRIPTION
**Pull request by rmetzger**
## What is the purpose of the change

Some e2e tests have the checking of logs for exceptions and errors disabled. When these tests fail, there is no way of getting the JM / TM logs.
With this change, these logs are included.

## Brief change log

- we register a on_exit hook / trap if we detect an azure environment which compresses the build artifact, when the build exists (for whatever reason)
- On error, and if ARTIFACTS_DIR is set, the test runner is copying the logs into ARTIFACTS_DIR.

### CODE REVIEW

File: flink-end-to-end-tests/run-nightly-tests.sh [hunk 1]
```diff
@@ -34,6 +34,23 @@ fi
...
 
+# On Azure CI, set artifacts dir
+if [ ! -z "$TF_BUILD" ] ; then
+	export ARTIFACTS_DIR="${END_TO_END_DIR}/artifacts"
+	mkdir -p $ARTIFACTS_DIR || { echo "FAILURE: cannot create log directory '${ARTIFACTS_DIR}'." ; exit 1; }
+
+	# compress and register logs for publication on exit
+	function compress_logs {
+		echo "COMPRESSING build artifacts."
+		ARTIFACTS_FILE=${BUILD_BUILDNUMBER}.tgz
```
Comment (zentol) [inline flink-end-to-end-tests/run-nightly-tests.sh]: can we name this `compressed_archive`? I keep getting confused and think this refers to a file within `ARTIFACTS_DIR`.

File: flink-end-to-end-tests/test-scripts/test-runner-common.sh [hunk 2]
```diff
@@ -74,12 +74,14 @@ function post_test_validation {
...
     fi
 
+    local failed=0
     # Investigate exit_code for failures of test executable as well as EXIT_CODE for failures of the test.
     # Do not clean up if either fails.
     if [[ ${exit_code} == 0 ]]; then
         if [[ ${EXIT_CODE} != 0 ]]; then
             printf "\n[FAIL] '${description}' failed after ${time_elapsed}! Test exited with exit code 0 but the logs contained errors, exceptions or non-empty .out files\n\n"
             exit_code=1
+            failed=1
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-end-to-end-tests/test-scripts/test-runner-common.sh]: Yep, you are right

File: flink-end-to-end-tests/run-nightly-tests.sh (hunk 1 above)
Comment (rmetzger) [reply flink-end-to-end-tests/run-nightly-tests.sh]: good idea! I was also confused a few times.

File: flink-end-to-end-tests/run-nightly-tests.sh [hunk 3]
```diff
@@ -206,17 +223,8 @@ printf "\n\n====================================================================
...
-HERE="`( cd \"${HERE}\" && pwd -P)`"
-if [ -z "${HERE}" ] ; then
-	# error; for some reason, the path is not accessible
-	# to the script (e.g. permissions re-evaled after suid)
-	exit 1  # fail
-fi
-ARTIFACTS_DIR="${HERE}/artifacts"
-mkdir -p $ARTIFACTS_DIR || { echo "FAILURE: cannot create log directory '${ARTIFACTS_DIR}'." ; exit 1; }
 
-LOG4J_PROPERTIES=${HERE}/../tools/log4j-travis.properties
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-end-to-end-tests/run-nightly-tests.sh]: Thanks. I will pay more attention next time.

File: flink-end-to-end-tests/run-nightly-tests.sh [hunk 4]
```diff
@@ -34,6 +34,23 @@ fi
...
+	export ARTIFACTS_DIR="${END_TO_END_DIR}/artifacts"
+	mkdir -p $ARTIFACTS_DIR || { echo "FAILURE: cannot create log directory '${ARTIFACTS_DIR}'." ; exit 1; }
+
+	# compress and register logs for publication on exit
+	function compress_logs {
+		echo "COMPRESSING build artifacts."
+		ARTIFACTS_FILE=${BUILD_BUILDNUMBER}.tgz
+		tar -zcvf ${ARTIFACTS_FILE} $ARTIFACTS_DIR
+		mkdir artifact-dir
+		cp ${ARTIFACTS_FILE} artifact-dir/
```
[the remark this answers is not in the document]
Comment (rmetzger) [reply flink-end-to-end-tests/run-nightly-tests.sh]: Yep. I will simplify this as much as I can.

**Comment by rmetzger** [comment]
Thank you for your review. I've incorporated all your feedback. I'm currently testing the changes in a private branch. I'll push them once they are stable.

**Comment by rmetzger** [comment]
I addressed your comments. This is ready for review again.

File: flink-end-to-end-tests/test-scripts/test-runner-common.sh [hunk 5]
```diff
@@ -94,6 +94,13 @@ function post_test_validation {
     if [[ ${exit_code} == 0 ]]; then
         cleanup
     else
+        # make logs available if ARTIFACTS_DIR is set
+        if [[ ${ARTIFACTS_DIR} != "" ]]; then
+            mkdir ${ARTIFACTS_DIR}/e2e-flink-logs 
+            cp $FLINK_DIR/log/* ${ARTIFACTS_DIR}/e2e-flink-logs/
```
Comment (zentol) [inline flink-end-to-end-tests/test-scripts/test-runner-common.sh]: nit: `mv` may be faster

**Comment by rmetzger** [comment]
Thanks a lot. I will merge it once the build has passed.

**Comment by zentol** [comment]
FYI, if you also want to upload the java e2e logs you can just set the `logBackupDir` property to `${ARTIFACTS_DIR}`.

**Comment by rmetzger** [comment]
Thanks! I pushed a commit adding this.

**Pull request closed without merging**